src/riscv32_csr_s.c

```diff
@@ -17,84 +17,111 @@ You should have received a copy of the GNU General Public License
+    csr_helper_masked(&vm->csr.status, dest, op, csr_sstatus_mask(vm));
+    riscv32_csr_isa_change(vm, PRIVILEGE_USER, cut_bits(vm->csr.status, CSR_STATUS_UXL_START, CSR_STATUS_UXL_SIZE));
     return true;
 }
 
-static bool riscv32_csr_sie(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_sie(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper_masked(&vm->csr.ie, dest, op, CSR_SEIP_MASK);
     return true;
 }
 
-static bool riscv32_csr_stvec(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_stvec(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper(&vm->csr.tvec[PRIVILEGE_SUPERVISOR], dest, op);
     return true;
 }
 
-static bool riscv32_csr_sscratch(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_sscratch(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper(&vm->csr.scratch[PRIVILEGE_SUPERVISOR], dest, op);
     return true;
 }
 
-static bool riscv32_csr_sepc(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_sepc(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper(&vm->csr.epc[PRIVILEGE_SUPERVISOR], dest, op);
     return true;
 }
 
-static bool riscv32_csr_scause(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_scause(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper(&vm->csr.cause[PRIVILEGE_SUPERVISOR], dest, op);
     return true;
 }
 
-static bool riscv32_csr_stval(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_stval(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper(&vm->csr.tval[PRIVILEGE_SUPERVISOR], dest, op);
     return true;
 }
 
-static bool riscv32_csr_sip(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_sip(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
     csr_helper_masked(&vm->csr.ip, dest, op, CSR_SEIP_MASK);
     return true;
 }
 
-static bool riscv32_csr_satp(riscv32_vm_state_t *vm, uint32_t csr_id, uint32_t* dest, uint8_t op)
+static bool riscv32_csr_satp(riscv32_vm_state_t *vm, uint32_t csr_id, reg_t* dest, uint8_t op)
 {
     UNUSED(csr_id);
-    uint32_t satp = (vm->mmu_virtual ? 0x80000000 : 0) | (vm->root_page_table >> 12);
+    reg_t satp;
+    bool is32bit = vm->isa[PRIVILEGE_SUPERVISOR] == ISA_RV32;
+    if (is32bit)
+    {
+	    satp = (!!vm->mmu_virtual << 31) | (vm->root_page_table >> 12);
+    }
+    else
+    {
+	    satp = ((reg_t)vm->mmu_virtual << 60) | ((vm->root_page_table >> 12) & gen_mask(44));
```

Comment:
GCC: warning: left shift count >= width of type
Probably should be separated with ifdef